File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/010-regular-expression-matching-approach2.py
from functools import lru_cache

def is_match(s, p):
    @lru_cache(maxsize=None)
    def match(i, j):
        if j == len(p):
            return i == len(s)
        first = i < len(s) and (p[j] == '.' or p[j] == s[i])
        if j + 1 < len(p) and p[j + 1] == '*':
            return match(i, j + 2) or (first and match(i + 1, j))
        return first and match(i + 1, j + 1)
    return match(0, 0)

assert is_match("aa", "a") == False
assert is_match("aa", "a*") == True
assert is_match("ab", ".*") == True
assert is_match("mississippi", "mis*is*p*.") == False
assert is_match("", "") == True
assert is_match("", "a*") == True
print("all tests pass")

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/010-regular-expression-matching-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function isMatch(s: string, p: string): boolean {
    const memo: Map<string, boolean> = new Map();
    function match(i: number, j: number): boolean {
        const key = `${i},${j}`;
        if (memo.has(key)) return memo.get(key)!;
        if (j === p.length) return i === s.length;
        const first = i < s.length && (p[j] === '.' || p[j] === s[i]);
        let result: boolean;
        if (j + 1 < p.length && p[j + 1] === '*') {
            result = match(i, j + 2) || (first && match(i + 1, j));
        } else {
            result = first && match(i + 1, j + 1);
        }
        memo.set(key, result);
        return result;
    }
    return match(0, 0);
}

assert(isMatch("aa", "a") === false);
assert(isMatch("aa", "a*") === true);
assert(isMatch("ab", ".*") === true);
assert(isMatch("mississippi", "mis*is*p*.") === false);
assert(isMatch("", "") === true);
assert(isMatch("", "a*") === true);
console.log("all tests pass");
